g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+into. Let's walk through the creation of the `toy.constant` operation:
 
-// Using a builder to create the operation and insert it where the builder
-// insertion point is currently set.
-Operation *customOperation = builder.createOperation(state);
+```mlir
+ %4 = "toy.constant"() {value = dense<1.0> : tensor<2x3xf64>} : () -> tensor<2x3xf64>
+```
 
-// An operation is not an SSA value (unlike LLVM), because it can return
-// multiple SSA values, the resulting value can be obtained:
-Value *result = customOperation->getResult(0);
+This operation takes zero operands, a
+[dense elements](../../LangRef.md#dense-elements-attribute) attribute named
+`value`, and returns a single result of
+[TensorType](../../LangRef.md#tensor-type). An operation inherits from the CRTP
+`mlir::Op` class which also takes some optional *traits* to customize its
+behavior. These traits may provide additional accessors, verification, etc.
+
+```c++
+class ConstantOp : public mlir::Op<ConstantOp,
+                      /// The ConstantOp takes zero inputs.
+                      mlir::OpTrait::ZeroOperands,
+                      /// The ConstantOp returns a single result.
+                      mlir::OpTrait::OneResult> {
+
+ public:
+  /// Inherit the constructors from the base Op class.
+  using Op::Op;
+
+  /// Provide the unique name for this operation. MLIR will use this to register
+  /// the operation and uniquely identify it throughout the system.
+  static llvm::StringRef getOperationName() { return "toy.constant"; }
+
+  /// Operations can provide additional verification beyond the traits they
+  /// define. Here we will ensure that the specific invariants of the constant
+  /// operation are upheld, for example the result type must be of TensorType.
+  LogicalResult verify();
+
+  /// Provide an interface to build this operation from a set of input values.
+  /// This interface is used by the builder to allow for easily generating
+  /// instances of this operation:
+  ///   mlir::OpBuilder::create<ConstantOp>(...)
+  /// This method populates the given `state` that MLIR uses to create
+  /// operations. This state is a collection of all of the discrete elements
+  /// that an operation may contain.
+  /// Build a constant with the given return type and `value` attribute.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::Type result, mlir::DenseElementsAttr value);
+  /// Build a constant and reuse the type from the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::DenseElementsAttr value);
+  /// Build a constant by broadcasting the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    double value);
+
+  /// Return the value of the constant by fetching it from the attribute.
+  mlir::DenseElementsAttr getValue();
+};
```

Comment:
This is too complex IMO: we need to build up concept and knowledge more slowly and progressively. That mean no need for more than strictly needed to understand the concept at hand (multiple builders don't seem needed here).

Reply:
I agree, I'd like to add a step-by-step intro to this in a followup.

Reply:
Beefed it up to include a more step-by-step introduction to the tablegen side.